From 4eace276aebc06138fe23afac4010eb519bab132 Mon Sep 17 00:00:00 2001
From: Jon Turney <jon.turney@dronecode.org.uk>
Date: Fri, 11 Jul 2014 23:37:02 +0100
Subject: [PATCH] Fix building minidump processor for MinGW

- Link minidump_stackwalk, minidump_dump, microdump_stackwalk with
PTHREAD_LIBS as pthread_cancel is used

- Link minidump_stackwalk, microdump_stackwalk, binarystream_unittest with
ws2_32 as ntoh functions are used

- _s function variants should only be used if _MSC_VER, use _r variants with
MinGW (and define _POSIX_C_SOURCE to ensure they are prototyped)

 - Don't try to build upload tools, dump_syms_dwarf for MinGW

(Part of https://breakpad.appspot.com/548002 was commited as svn r1399 to
fix breakpad client compilation for MinGW

This is mainly the configure.ac and Makefile.am changes left over from
https://breakpad.appspot.com/548002/ with a bit of updating, to fix building
of the minidump processor library and tools for MinGW)

v2:
Use _POSIX_C_SOURCE not _POSIX to enable *_r() prototypes in headers

The headers seem to have changed so localtime_r() prototype is now guarded
by _POSIX_C_SOURCE

This seems right anyhow as _POSIX_C_SOURCE is what SUS defines, whereas I
think _POSIX is some left-over Interix thing?

v3:
Drop _snprintf fix, now dealt with centrally and correctly since 48673cdb

Signed-off-by: Jon Turney <jon.turney@dronecode.org.uk>
---
 Makefile.am  | 15 ++++++++++++---
 configure.ac |  6 ++++++
 2 files changed, 18 insertions(+), 3 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 702a7a87..74497797 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -55,6 +55,10 @@ if WINDOWS_HOST
 AM_CPPFLAGS += -DNO_STABS_SUPPORT
 endif
 
+if MINGW_HOST
+SOCKET_LIBS = -lws2_32
+endif
+
 # Specify include paths for ac macros
 ACLOCAL_AMFLAGS = -I m4
 
@@ -897,6 +901,7 @@ src_tools_linux_md2core_minidump_2_core_unittest_LDADD = \
 	$(PTHREAD_CFLAGS) $(PTHREAD_LIBS)
 
 if WINDOWS_HOST
+if !MINGW_HOST
 if !DISABLE_TOOLS
 src_tools_windows_dump_syms_dwarf_dump_syms_SOURCES = \
 	src/common/dwarf_cfi_to_module.cc \
@@ -914,6 +919,7 @@ src_tools_windows_dump_syms_dwarf_dump_syms_SOURCES = \
 	src/tools/windows/dump_syms_dwarf/dump_syms.cc
 endif
 endif
+endif
 
 src_processor_address_map_unittest_SOURCES = \
 	src/processor/address_map_unittest.cc
@@ -1458,7 +1464,8 @@ src_processor_minidump_dump_LDADD = \
 	src/processor/logging.o \
 	src/processor/minidump.o \
 	src/processor/pathname_stripper.o \
-	src/processor/proc_maps_linux.o
+	src/processor/proc_maps_linux.o \
+	$(PTHREAD_CFLAGS) $(PTHREAD_LIBS)
 
 src_processor_microdump_stackwalk_SOURCES = \
 	src/processor/microdump_stackwalk.cc
@@ -1495,7 +1502,8 @@ src_processor_microdump_stackwalk_LDADD = \
 	src/processor/stackwalker_sparc.o \
 	src/processor/stackwalker_x86.o \
 	src/processor/tokenize.o \
-	src/third_party/libdisasm/libdisasm.a
+	src/third_party/libdisasm/libdisasm.a \
+	$(PTHREAD_CFLAGS) $(PTHREAD_LIBS) $(SOCKET_LIBS)
 if LINUX_HOST
 src_processor_microdump_stackwalk_LDADD += \
 	src/common/linux/scoped_pipe.o \
@@ -1543,7 +1551,8 @@ src_processor_minidump_stackwalk_LDADD = \
 	src/processor/stackwalker_x86.o \
 	src/processor/symbolic_constants_win.o \
 	src/processor/tokenize.o \
-	src/third_party/libdisasm/libdisasm.a
+	src/third_party/libdisasm/libdisasm.a \
+	$(PTHREAD_CFLAGS) $(PTHREAD_LIBS) $(SOCKET_LIBS)
 if LINUX_HOST
 src_processor_minidump_stackwalk_LDADD += \
 	src/common/linux/scoped_pipe.o \
diff --git a/configure.ac b/configure.ac
index 368e517a..53be0ad4 100644
--- a/configure.ac
+++ b/configure.ac
@@ -104,10 +104,16 @@ case $host in
   *-*-cygwin* )
     WINDOWS_HOST=true
     ;;
+  *-*-mingw* )
+    WINDOWS_HOST=true
+    MINGW_HOST=true
+    ;;
 esac
 
 AM_CONDITIONAL(LINUX_HOST, test x$LINUX_HOST = xtrue)
+# WINDOWS_HOST means MINGW or CYGWIN
 AM_CONDITIONAL(WINDOWS_HOST, test x$WINDOWS_HOST = xtrue)
+AM_CONDITIONAL(MINGW_HOST, test x$MINGW_HOST = xtrue)
 
 # Only use Android support headers when compiling for Android
 case $host in
-- 
2.53.0.windows.1

